Unix/http/httpauth.c

```diff
@@ -0,0 +1,1400 @@
+
+        encode_context.size = header_start_len + protocol_len + 1;
+
+        // Call again to copy the data
+        if (rslt >= 0) {
+            rslt = Base64Enc(token->value, token->length, EncodePlaceCallback, &encode_context);
+        }
+
+        memcpy(encode_context.pdata + encode_context.size, HEADER_TAIL, HEADER_TAIL_LEN);
+        encode_context.size += HEADER_TAIL_LEN;
+        encode_context.pdata[encode_context.size] = '\0';
+
+    }
+
+    if (pResultLen) {
+        *pResultLen = encode_context.size;
+    }
+    return encode_context.pdata;
+}
+
+#endif
+
+MI_Boolean IsClientAuthorized(_In_ Http_SR_SocketData * handler)
+{
+    MI_Boolean authorised = FALSE;
+    HttpHeaders *headers = &handler->recvHeaders;
+    static const char *RESPONSE_HEADER_UNAUTH_FMT =
+        "HTTP/1.1 401 Unauthorized\r\n" "Content-Length: 0\r\n"
+        "WWW-Authentication: Basic realm=\"WSMAN\"\r\n" "WWW-Authentication: Negotiate\r\n" "\r\n";
+
+#if AUTHORIZATION
+    static const char *RESPONSE_HEADER_BAD_REQUEST =
+        "HTTP/1.1 400 Bad Request\r\n" "Content-Length: 0\r\n" "\r\n";
+    OM_uint32 flags = 0;
+    const char *protocol_p = NULL;
+#endif
+
+    unsigned char *auth_response = NULL;
+    int response_len = 0;
+
+    if (IsAuthCallsIgnored()) {
+        handler->httpErrorCode = 0; // We let the transaction set the error code
+        handler->isAuthorised = TRUE;
+        return TRUE;
+    }
+
+    if (!headers) {
+        handler->httpErrorCode = HTTP_ERROR_CODE_INTERNAL_SERVER_ERROR;
+        goto Done;
+    }
+
+    if (Strncasecmp(headers->authorization, AUTHENTICATION_BASIC, AUTHENTICATION_BASIC_LENGTH) == 0) {
+        if (!headers->username || !headers->password
+            || 0 != AuthenticateUser(headers->username, headers->password)) {
+            handler->httpErrorCode = HTTP_ERROR_CODE_UNAUTHORIZED;
+            auth_response = (unsigned char *)RESPONSE_HEADER_UNAUTH_FMT;
+            response_len = strlen(RESPONSE_HEADER_UNAUTH_FMT);
+
+
+            trace_HTTP_UserAuthFailed("user not authenticated");
+            handler->authFailed = TRUE;
+
+            _SendAuthResponse(handler, auth_response, response_len);
+            return FALSE;
+        }
+
+        if (0 != LookupUser(headers->username, &handler->authInfo.uid, &handler->authInfo.gid)) {
+            trace_GetUserUidGid_Failed(headers->username);
+
+            handler->httpErrorCode = HTTP_ERROR_CODE_UNAUTHORIZED;
+            auth_response = (unsigned char *)RESPONSE_HEADER_UNAUTH_FMT;
+            response_len = strlen(RESPONSE_HEADER_UNAUTH_FMT);
+
+            trace_HTTP_UserAuthFailed("basic auth user creds not present");
+            handler->authFailed = TRUE;
+            _SendAuthResponse(handler, auth_response, response_len);
+            return FALSE;
+        }
+
+        handler->httpErrorCode = 0; // Let the request do the error code
+        handler->isAuthorised = TRUE;
+        return TRUE;
+    }
+#ifdef AUTHORIZATION
+    else {
```

Comment:
This isn't necessarily negotiate. We should check what happens with another auth type.